### ISSUE TITLE
Custom loss for LightGBM lead to negative probability prediction

### ISSUE THREAD

**Issue by ahalaha** [microsoft/LightGBM#6389]
## Description
I am trying to do custom loss for binary classification model using LightGBM, but the trained model prediction returned negative probabilities.

## Reproducible example
I follow this blog https://maxhalford.github.io/blog/lightgbm-focal-loss/ to implement Focal Loss on LightGBM binary classifier.
```
class FocalLoss:

    def __init__(self, gamma, alpha=None):
        self.alpha = alpha
        self.gamma = gamma

    def at(self, y):
        if self.alpha is None:
            return np.ones_like(y)
        return np.where(y, self.alpha, 1 - self.alpha)

    def pt(self, y, p):
        p = np.clip(p, 1e-15, 1 - 1e-15)
        return np.where(y, p, 1 - p)

    def __call__(self, y_true, y_pred):
        at = self.at(y_true)
        pt = self.pt(y_true, y_pred)
        return -at * (1 - pt) ** self.gamma * np.log(pt)

    def grad(self, y_true, y_pred):
        y = 2 * y_true - 1  # {0, 1} -> {-1, 1}
        at = self.at(y_true)
        pt = self.pt(y_true, y_pred)
        g = self.gamma
        return at * y * (1 - pt) ** g * (g * pt * np.log(pt) + pt - 1)

    def hess(self, y_true, y_pred):
        y = 2 * y_true - 1  # {0, 1} -> {-1, 1}
        at = self.at(y_true)
        pt = self.pt(y_true, y_pred)
        g = self.gamma

        u = at * y * (1 - pt) ** g
        du = -at * y * g * (1 - pt) ** (g - 1)
        v = g * pt * np.log(pt) + pt - 1
        dv = g * np.log(pt) + g + 1

        return (du * v + u * dv) * y * (pt * (1 - pt))

    def init_score(self, y_true):
        res = optimize.minimize_scalar(
            lambda p: self(y_true, p).sum(),
            bounds=(0, 1),
            method='bounded'
        )
        p = res.x
        log_odds = np.log(p / (1 - p))
        return log_odds

    def lgb_obj(self, preds, dtrain):
        y_true = dtrain.get_label()
        p = special.expit(preds)
        return self.grad(y_true, p), self.hess(y_true, p)

    def lgb_eval(self, preds, dtrain):
        y_true = dtrain.get_label()
        p = special.expit(preds)
        is_higher_better = False
        return 'focal_loss', self(y_true, p).mean(), is_higher_better
```

Then just a simple classifier:

```
import lightgbm as lgb
from sklearn.datasets import load_breast_cancer
from sklearn.model_selection import train_test_split


dataset = load_breast_cancer(as_frame=True)

X_train, X_test, y_train, y_test = train_test_split(dataset.data, dataset.target , test_size=0.2, random_state=0)

x_tr, x_vl, y_tr, y_vl = train_test_split(X_train,y_train,test_size = 0.2, random_state=0)


fl = FocalLoss(alpha=0.2, gamma=2)

focal_loss = fl.lgb_obj
eval_error = fl.lgb_eval

params = {
    "boosting_type": "gbdt",
    "objective": focal_loss,
    "metric": ["auc","rmse"],
    "verbose": 10,
    "random_state": 42,
    "device": "cpu",
}

d_train = lgb.Dataset(x_tr, y_tr)
d_valid = lgb.Dataset(x_vl, y_vl,reference=d_train)

print('Start training')
model = lgb.train(params=params,train_set=d_train,
                  valid_sets=[d_train, d_valid],
                  callbacks=[lgb.early_stopping(stopping_rounds=100,verbose=10),],
                  feval=eval_error
                  )
```
The model is definitely bad, but the problem is when I call the predict method, this is what on return:

```
model.predict(X_test)

array([-0.50880468,  0.3693605 ,  0.52534365,  0.31979503,  0.46992679,
        0.52635717,  0.42439338,  0.52743569,  0.50436045,  0.47013601,
       -0.45913891,  0.21876511,  0.53404664, -0.12413401, -0.43392308,
       -0.44023844,  0.18890394, -0.5244376 , -0.52448758, -0.52448758,
       -0.5100286 , -0.51785048,  0.37074254,  0.52573731, -0.44023844,
        0.52431002,  0.52461526, -0.44361862,  0.52430158, -0.53164517,
        0.52442738, -0.47865461,  0.34670623, -0.51634395,  0.52529346,
       -0.51952488,  0.52523917, -0.44351197,  0.31979503, -0.52792125,
       -0.04178972,  0.29425045,  0.1141792 ,  0.5250677 , -0.11963795,
       -0.52446354,  0.52548207,  0.44207354,  0.52504278, -0.52590175,
       -0.52449109, -0.50045889, -0.5244715 ,  0.52483863,  0.536903  ,
        0.5254319 ,  0.3693605 ,  0.28750883,  0.3693605 , -0.52455554,
       -0.50880468, -0.52613068,  0.3553297 ,  0.52446307, -0.52451552,
       -0.11313199, -0.52484974, -0.5244002 , -0.52449554,  0.50211657,
       -0.04761826, -0.52534093,  0.52424519, -0.3558752 , -0.48744658,
        0.47222388,  0.52584841,  0.37074254,  0.52469911,  0.29683422,
       -0.51634395, -0.5245152 , -0.52548014,  0.47179732, -0.44023844,
        0.52425657,  0.52471657,  0.52539643, -0.51776054, -0.5245152 ,
        0.52933499, -0.12996255, -0.221874  , -0.51755596,  0.35735785,
        0.40885033, -0.52473433, -0.06279767,  0.4737567 ,  0.49707636,
        0.35572469,  0.52493403,  0.31979503,  0.27942182, -0.52482029,
        0.3580971 , -0.52453118,  0.26388491, -0.07564908, -0.49671874,
        0.50110396, -0.52552423, -0.52447412,  0.19961468])
```
There is negative probability, and that makes no sense to me. These probabilities should be in range (0, 1).
Anyone please tells me what I'm doing wrong.

## Environment info

I'm currently using LightGBM 4.2.0 via

```
pip install lightgbm

```

**Comment by jmoralez** [comment]
Hey @ahalaha, thanks for using LightGBM. When you use a custom objective the output of predict is always the raw score, so you have to apply sigmoid to transform that into probabilities (note that's what is done in the eval function `p = special.expit(preds)`).
